### DISCUSSION TITLE
Replace built-in icons with SVG?

### DISCUSSION THREAD

**Discussion by MarcinOrlowski** [logisim-evolution/logisim-evolution#640] [Ideas]
Are there any objections to completely get rid of built-in "coded" icons and replace them with SVG (i.e. via Apache Batik or anything lighter)? That'd remove lot of useless code and allow support for different icon sets (i.e. light/dark). Current icons are both ugly and hardcoded, which makes a bit problematic to properly deal with darl/light themes etc. Not to mention animated icons that are (IMHO) confusing and pretty bad UX.

**Comment by BFH-ktt1** [comment]
Yes, for following reasoning:
Logisim was build up with icons of size 16x16 pixels, and in the original code all icons were .png files. As we build in the automatic scaling of logisim-evolution, the scaling of the icons was awful, hence all icons were replaced by scalable icons. Although svg icons might solve this problem it does not solve the problem of being visible, as in svg-files the colors are also "hardcoded". Hence icons that look nice on dark background will not be visible on light or visa-versa. Furthermore, they might not be visible at all when scaling is on 1 (hence icons of 16x16 pixels).

To solve the problem of visibility (which I think is a problem), I see more in the direction to replace the colors in the icons from fixed to taking the background color of the theme (which can be done easily) and xor it with 0xFFFFFF, hence for a light theme it becomes dark, and for a dark theme it becomes light.

About "ugly", this is in the eye of the beholder :-)

As for the animated icons, this we can supres without a problem, was just a "fun" part that I build in once :-)

**Reply by MarcinOrlowski** [reply]
I think that having icons as separate SVG files will solve visibility problem too. Not automatically, but by having separate set of dark and light icons user would be able to switch between. And while icons would be external, one can even come with own set by their liking or design.

As for "ugly", well, true :) But I rather meant that current ones are clearly made by someone not being a gfx designer, so I see the possibility for the improvements once the possibility gets opened :)

**Reply by maehne** [reply]
I am not opposed to replacing the hard-coded icons with SVG icons, as long as the icons look equally good on low DPI and high DPI screens. Personally, I find the animated icons very distracting and therefore turn them off systematically. Using SVG icons is standard for wide-spread desktop environments on Linux, which typically offer a broad choice of themes. Indeed, good programmers are not necessarily good graphics designers, too -- and vice versa. Therefore, it might help attracting contributions from such a side.

The hard-coded icons have the advantage that you can relatively simply adapt them to different color themes if we ensure that the used colours are not hard-coded but configurable. This ensures consistency and would give users the choice to adapt the colours to their liking. A suitable default could be defined programmatically based on the colours of the chosen Look and Feel. 

A dedicated SVG icon set would probably need to be defined for each Look and Feel using different base colours unless there's a way to use some colour palette, which can be externally adjusted to the one used by the Look and Feel. Another risk is an inconsistency between the SVG icons for components and their look on a schematic sheet. One cause may be the selected gate shape (ANSI or IEC).

**Comment by MarcinOrlowski** [comment]
I did some preliminary tests using Batik. It is pretty simple to be incorporated as it comes with Swing component, but I cannot call I was successful as more advanced designs I made in Inkscape did not work as expected (left expected, right actual):

![test](https://user-images.githubusercontent.com/8041294/121784553-cedbdf80-cbb4-11eb-9123-c60b723dfa40.png)

Not sure what I did wrong yet, especially: second gate's gradient did not work, but first one's is drawn. Some plain horizontal lines are gone while others (same thickness) are here, colors are not really as expected, main text's glow is gone, etc. Nonetheless, for basic shapes like these in i. e. toolbar can be rendered correctly right now, that's the good step forward anyway.

**Reply by MarcinOrlowski** [reply]
Still not sure why certain filters (i.e. shadows) do not work with Batik. Maybe I am over-interpreting the docs or maybe these are not supported, or maybe Inkscape writes stuff differently from what Batik expects. Interesting thing is that even web browsers render this file correctly, but not Batik. Still this is not a block-stopper really as icons do not need shadows at the start anyway.

**Comment by mbaillif** [comment]
Hello,
The SVG format is excellent, but it requires more work for the realization and it is more complicated to integrate with the Html of the documentation.

Independently of that
I prepared for the doc some 64x64 / 32x32 icons in PNG which also allows transparency.  
The software adjusts them quite well in smaller size. 
This is to make them look better on 4k screens for example.
These icons could be used in the software with a downward adjustment instead of upward.
Available on  
mbaillif /logisim-evolution 
in the "ressources\doc\icones" folder

To be experimented. 

I will also prepare some SVG icons to test.

**Reply by MarcinOrlowski** [reply]
Yep, I already figured out I was far too optimistic with SVG for icons.

**Reply by maehne** [reply]
One interesting aspect of using SVG for the icons would be styling the icons with a LAF-theme-dependent CSS to choose colours with good contrast to the background. Would Batik support that?

@mbaillif: Thanks for preparing higher resolution icons for the documentation!

**Comment by mbaillif** [comment]
The test icons of the lib wiring in svg are in \doc\ icons\svgwithbackground
and svgwithoutbackground dernier pull_request

To give precisions on my previous statement.
SVG are easy to integrate in Html5 but not in javadoc which is very obsolete.
Then when you have a svg it's easy to make an icon if needed.

I'm a big supporter of svg
